### PR TITLE
[SPARK-26382][CORE] prefix comparator should handle -0.0

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

This is kind of a followup of https://github.com/apache/spark/pull/23239

The `UnsafeProject` will normalize special float/double values(NaN and -0.0), so the sorter doesn't have to handle it.

However, for consistency and future-proof, this PR proposes to normalize `-0.0` in the prefix comparator, so that it's same with the normal ordering. Note that prefix comparator handles NaN as well.

This is not a bug fix, but a safe guard.

## How was this patch tested?

existing tests

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @dongjoon-hyun @viirya @kiszk @adoron

File: core/src/test/scala/org/apache/spark/util/collection/unsafe/sort/PrefixComparatorsSuite.scala [hunk 1]
```diff
@@ -134,22 +135,32 @@ class PrefixComparatorsSuite extends SparkFunSuite with PropertyChecks {
     assert(java.lang.Double.doubleToRawLongBits(negativeNan) < 0)
     val prefix = PrefixComparators.DoublePrefixComparator.computePrefix(negativeNan)
     val doubleMaxPrefix = PrefixComparators.DoublePrefixComparator.computePrefix(Double.MaxValue)
+    // -NaN is greater than the max double value.
     assert(PrefixComparators.DOUBLE.compare(prefix, doubleMaxPrefix) === 1)
   }
 
   test("double prefix comparator handles other special values properly") {
-    val nullValue = 0L
+    // See `SortPrefix.nulValue` for how we deal with nulls for float/double type
```
Comment (cloud-fan) [inline core/src/test/scala/org/apache/spark/util/collection/unsafe/sort/PrefixComparatorsSuite.scala]: improve the test coverage a little bit.

**Review by viirya — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

**Comment by cloud-fan** [comment]
@viirya yes you are right

**Comment by kiszk** [comment]
LGTM pending jenkins
nit: For our future, it would be good to add a link to this PR from bullet 1 in the description of #23239

File: core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PrefixComparators.java [hunk 2]
```diff
@@ -69,6 +69,8 @@ public static long computePrefix(byte[] bytes) {
      * details see http://stereopsis.com/radix.html.
      */
     public static long computePrefix(double value) {
+      // normalize -0.0 to 0.0, as they should be equal
+      value = value == -0.0 ? 0.0 : value;
```
Comment (dongjoon-hyun) [inline core/src/main/java/org/apache/spark/util/collection/unsafe/sort/PrefixComparators.java]: Technically, `value == -0.0` is `true` for both `0.0` and `-0.0`, but the current one is the simplest one for this normalization. We don't need to exclude `0.0` here. +1 for the logic.

**Comment by dongjoon-hyun** [comment]
Shall we use `[CORE]` instead of `[SQL]` since this touches only `PrefixComparators` in `core`?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by dongjoon-hyun** [comment]
Although this is not a bug fix, can we have this in `branch-2.4`, @cloud-fan and @gatorsmile ? I'm +1 for backporting.

**Comment by cloud-fan** [comment]
@kiszk good idea, updated #23239

@dongjoon-hyun I'm fine with it. Feel free to merge :)

**Comment by dongjoon-hyun** [comment]
Thank you, @cloud-fan and @viirya , @HyukjinKwon , @kiszk .
Merged to master/branch-2.4 .

**Pull request closed without merging**